
### Contents List
 
#### 01 Importing the libraries
#### 02 Importing the dataframes
#### 03 Finding and fixing mixed-type data
#### 04 Finding Missing Values
#### 05 Addressing Missing Values
#### 06 Finding Duplicates
#### 07 Addressing Duplicates
#### 08 Task 4.5
- 8-1 Question 2.
  - 8-1-2. Checking "df_ords" dataframe.
- 8-2 Question 3. Checking for mixed-type data in "df_ords" dataframe
- 8-3 Question 5. Running a check for missing values in "df_ords" dataframe.
- 8-4 Question 6. Addressing the missing values using an appropriate method
- 8-5 Question 7. Running a check for duplicate values in "df_ords" data.
- 8-6 Question 9. Exporting the final, cleaned "df_prods" and "df_ords data" as “.csv” in “Prepared Data”
** **

# 01 Importing the libraries

In [40]:
#importing the libraries
import pandas as pd
import numpy as np
import os

# 02 Importing the dataframes

In [41]:
#importing the “orders.csv” and “products.csv" dataframes via "path" string shortcut (reuseable path variable)
path= r'C:\Users\mehrn\OneDrive\Desktop\CareerFoundry Materials\Tasks\Python\10-01-23 Instacart Basket Analysis'
df_prods = pd.read_csv (os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

# 03 Finding and fixing mixed-type data

In [91]:
# Create a test dataframe with mixed type column 
df_test = pd.DataFrame()
df_test['mix'] =['a', 'b', 1 , True, np.nan]
df_test.head()

mix
0     a
1     b
2     1
3  True
4   NaN

In [43]:
# The function for checking whether a dataframe contains any mixed-type columns
# Within the for-loop, a new variable is created: weird. Assigned to it is a test that checks whether 
# the data types within the column are consistent. The weird variable will ultimately take a boolean 
# value of either True or False. If True, that means the column contains inconsistent data types. 
# If False, that means the column contains only one data type. Boolean values can also be represented 
# by numbers: 0 as False and 1 as True.
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
# the if statement is checking whether weird is true or false. If it’s greater than 0, than it’s true. If not, 
# it’s false. If weird is true, the command print(col) is executed, which prints the problematic column for you 
# to see. Because of the for-loop, this command will be executed on every column in your dataframe, printing every 
# mixed-type column it finds.
  if len (df_test[weird]) > 0:
    print (col)

mix


In [44]:
#Based on most values or knowledge of analysis we would change the datatype
# To change this, simply update the str within the astype() function to int64 or whichever numeric data type you want to use
df_test['mix'] = df_test['mix'].astype('str')

# 04 Finding Missing Values


In [80]:
df_test.isnull().sum()

mix    0
dtype: int64

In [45]:
# sum() function obtains the total number of missing observations. True values interpreted numerically as 1, and False as 0.
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [46]:
# subsetting df_prods to df_nan to see the 16 null values. We created a subset of the dataframe containing null product_name
df_nan = df_prods [df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# 05 Addressing Missing Values

In [47]:
#There are a few ways to deal with missing data:

# Create a new variable that acts like a flag based on the missing value.
# Impute the value with the mean or median of the column (if the variable is numeric).
# Remove or filter out the missing data

#If you choose to use the median, you can find it using the df_prods.median() function,

#df['column with missings'].fillna(mean value, inplace=True)


#If you choose to use the mean, you can use the df.describe() function to find the mean and replace.

#df['column with missings'].fillna(median value, inplace=True)

In [48]:
# filtering out (the ones that aren’t missing their "product_name") into a subset "df_prods"
df_prods_clean= df_prods[df_prods ['product_name'].isnull() == False]
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [49]:
df_prods.shape

(49693, 5)

In [92]:
# Another way you can drop all missing values is via the following command:
df_test.dropna(inplace= False)
#df_prods.dropna(inplace = False)
df_test

mix
0     a
1     b
2     1
3  True
4   NaN

In [94]:
# If you wanted to use this command to drop only the NaNs from a particular column, the code would look like this:

# df_prods.dropna(subset = [‘product_name’], inplace = True)
df_test.dropna(subset = ['mix'], inplace= True)
df_test

mix
0     a
1     b
2     1
3  True

# 06 Finding Duplicates

In [52]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [53]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# 07 Addressing Duplicates

In [54]:
# Deleting the identified duplicate rows
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [55]:
df_prods_clean_no_dups.shape

(49672, 5)

# 08 Task 4.5

### 8-1 Question 2. Run the df.describe() function on your df_ords dataframe. Share in a markdown cell whether anything about the data looks off or should be investigated further.

In [56]:
# Running the describe() function on "df_ords" dataframe
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### Result: All the values in "df_ordes" columns seem to be consistent based on the descriptive statistics.

### 8-1-2. Checking "df_ords" dataframe.

In [57]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### Result: maximun price in prices, 99999, is too high in "df_prods" and seems to be an outlier for a grocery item (milk 2%).

In [58]:
# Finding the product with the price of 99999
df_prods[df_prods.prices == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

### 8-2 Question 3. Checking for mixed-type data in "df_ords" dataframe

In [59]:
# Checking for the mixed datatypes in a df_ords column
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
  else: 
    print(col, ' is concistent')

order_id  is concistent
user_id  is concistent
eval_set  is concistent
order_number  is concistent
order_dow  is concistent
order_hour_of_day  is concistent
days_since_prior_order  is concistent


#### Result:  No mixed-type column found in "df_ords"

### 8-3 Question 5. Running a check for missing values in "df_ords" dataframe.

In [60]:
# Checking for missing values in "df_ords" dataframe
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### Result: There are 206209 missing value for "days_since_prior_order" column which might be because of the orders belong to the first-time buyers. To make sure, we assumed that each user_id (unique user) in "ords_df" has a first transaction. Then the count of distinct users must be the same as the sum of missing values:

In [61]:
# Counting distinct number of users
df_ords['user_id'].nunique()

206209

#### Result: The same number as missing values came up as the distinct number of users which proves the missing values are because of useres being the first-time buyer. To address the missing values, we would either impute the "Nan" values to "-1" or create a new variable that acts like a flag (boolean) to distinguish between new and old customers .

### 8-4 Question 6. Addressing the missing values using an appropriate method

In [62]:
# Creating a dataframe for missing values
df_ords_nan = df_ords [df_ords ['days_since_prior_order'].isnull() == True]
df_ords_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [63]:
# Imputing missing values with a negetive number(-1)
df_ords['days_since_prior_order'].fillna(-1, inplace = True)
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                    -1.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [64]:
# Creating a Flag variable for new customers to address the missing values
df_ords_clean = df_ords
df_ords_clean ['new_customer'] = df_ords['days_since_prior_order'] == -1
df_ords_clean

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                    -1.0          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 8 columns]

### 8-5 Question 7. Running a check for duplicate values in "df_ords" data.

In [95]:
# Checking for duplicate values in "df_ords"
df_dups = df_ords_clean[df_ords_clean.duplicated()]
df_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

#### Result: There is no duplicate values in "df_ords_clean"

### 8-6 Question 9. Exporting the final, cleaned "df_prods" and "df_ords data" as “.csv” in “Prepared Data”

In [66]:
#Exporting 'df_ords_clean' to "orders_checked_csv"
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.CSV'))

In [67]:
#Exporting 'df_prods_clean_no_dups' to "orders_checked_csv"
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products__checked.CSV'))